File: src/game/progression.ts
import { player, state } from "../state";
import { pulseText } from "../entities/particles";
import { xpToNextLevel } from "./balance";
import { markHudDirty, markUpgradeReady } from "../simulation/events";

export function collectExperience(amount: number): void {
  state.xp += amount;
  state.score += amount * 3;
  pulseText(player.x, player.y - 34, `+${amount} XP`, "#72ffb1");
  markHudDirty();

  while (state.xp >= state.xpTarget) {
    state.xp -= state.xpTarget;
    state.level += 1;
    state.xpTarget = xpToNextLevel(state.level);
    state.pendingUpgrades += 1;
    pulseText(player.x, player.y - 58, `Niveau ${state.level}`, "#ffbf47");
    markUpgradeReady();
  }
}

File: sim/crates/voidline-sim/src/progression.rs
//! XP collection + level-up, mirroring `src/game/progression.ts:collectExperience`.

use voidline_data::balance::Balance;

use crate::balance_curves::xp_to_next_level;
use crate::player::Player;
use crate::state::GameState;

pub fn collect_experience(
    balance: &Balance,
    state: &mut GameState,
    player: &mut Player,
    amount: u32,
) {
    state.xp += amount;
    state.score += (amount as f64) * 3.0;
    if player.traits.magnet_storm {
        player.magnet_storm_charge = (player.magnet_storm_charge + amount as f64).min(80.0);
    }

    while state.xp >= state.xp_target {
        state.xp -= state.xp_target;
        state.level += 1;
        state.xp_target = xp_to_next_level(balance, state.level) as u32;
        state.pending_upgrades += 1;
    }
}
